### PR TITLE
Enable limit on range Of JSON-RPC API trace_filter method (#5971)

### LINKED ISSUE #5971: The JSON-RPC trace_filter method can cause a Thread block
### Description
Hi, while I was testing the JSON-RPC API I noticed that the _filter_ method of _TRACE_ API accepts a large block range which can cause a Thread block exception and possible leads to DoS the node.

I'm receiving the exception when sending a block range like 0 - 100000000

### Steps to Reproduce (Bug)
1. Enable the JSON-RPC and TRACE API on the Node
2. Call the trace_filter method with a large block range
3. curl -X POST --data '{"jsonrpc":"2.0","method":"trace_filter","params":[{"fromBlock":"1","toBlock":"1000000000","after":2,"count":2,"fromAddress":["0xfe3b557e8fb62b89f4916b721be55ceb828dbd73"]}],"id":415}' http://127.0.0.1:8548
4. Checking the logs, there will be a lot of Thread blocked exceptions, which will need to reboot the node to stop it from spawn

**Expected behavior:** An error message when the supplied block range is too large

**Actual behavior:** The node will fetch the traces between the entire block range causing a Thread exception, and keeping the client waiting until timeout the request

**Frequency:** Every time when a large block range is supplied.

#### Short Fix

A limit for the block range can be applied as occurred on the [eth_getLogs](https://besu.hyperledger.org/23.4.0/public-networks/reference/api#eth_getlogs)

### Logs (if a bug)
```
2023-09-28 15:49:54.013-03:00 | BftProcessorExecutor-QBFT-0 | INFO  | QbftBesuControllerBuilder | Imported #16,997 / 0 tx / 0 pending / 0 (0.0%) gas / (0xf4749789dab086b9701de03979fab0bb79809b3fdebf7300b78e67d07298a01c)
2023-09-28 15:49:54.013-03:00 | BftProcessorExecutor-QBFT-0 | INFO  | QbftBesuControllerBuilder | Produced #16,997 / 0 tx / 0 pending / 0 (0.0%) gas / (0xf4749789dab086b9701de03979fab0bb79809b3fdebf7300b78e67d07298a01c)
2023-09-28 15:49:54.013-03:00 | BftProcessorExecutor-QBFT-0 | INFO  | QbftBesuControllerBuilder | Imported #16,997 / 0 tx / 0 pending / 0 (0.0%) gas / (0xf4749789dab086b9701de03979fab0bb79809b3fdebf7300b78e67d07298a01c)
2023-09-28 15:49:54.015-03:00 | BftProcessorExecutor-QBFT-0 | INFO  | QbftBesuControllerBuilder | Imported #16,997 / 0 tx / 0 pending / 0 (0.0%) gas / (0xf4749789dab086b9701de03979fab0bb79809b3fdebf7300b78e67d07298a01c)
2023-09-28 15:49:54.037-03:00 | vertx-blocked-thread-checker | WARN  | BlockedThreadChecker | Thread Thread[#67,vert.x-worker-thread-17,5,main] has been blocked for 62529 ms, time limit is 60000 ms
io.vertx.core.VertxException: Thread blocked
	at org.rocksdb.RocksDB.get(Native Method)
	at org.rocksdb.RocksDB.get(RocksDB.java:2096)
	at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage.get(RocksDBColumnarKeyValueStorage.java:253)
	at org.hyperledger.besu.plugin.services.storage.rocksdb.segmented.RocksDBColumnarKeyValueStorage.get(RocksDBColumnarKeyValueStorage.java:69)
	at org.hyperledger.besu.services.kvstore.SegmentedKeyValueStorageAdapter.get(SegmentedKeyValueStorageAdapter.java:70)
	at org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStoragePrefixedKeyBlockchainStorage.get(KeyValueStoragePrefixedKeyBlockchainStorage.java:140)
	at org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStoragePrefixedKeyBlockchainStorage.getBlockHash(KeyValueStoragePrefixedKeyBlockchainStorage.java:112)
	at org.hyperledger.besu.ethereum.chain.DefaultBlockchain.getBlockHashByNumber(DefaultBlockchain.java:250)
	at org.hyperledger.besu.ethereum.chain.Blockchain.getBlockByNumber(Blockchain.java:108)
	at org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.TraceFilter.getBlockList(TraceFilter.java:207)
	at org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.TraceFilter.traceFilterWithPipeline(TraceFilter.java:124)
	at org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.TraceFilter.response(TraceFilter.java:99)
	at org.hyperledger.besu.ethereum.api.jsonrpc.execution.BaseJsonRpcProcessor.process(BaseJsonRpcProcessor.java:44)
	at org.hyperledger.besu.ethereum.api.jsonrpc.execution.TracedJsonRpcProcessor.process(TracedJsonRpcProcessor.java:41)
	at org.hyperledger.besu.ethereum.api.jsonrpc.execution.TimedJsonRpcProcessor.process(TimedJsonRpcProcessor.java:45)
	at org.hyperledger.besu.ethereum.api.jsonrpc.execution.JsonRpcExecutor.execute(JsonRpcExecutor.java:92)
	at org.hyperledger.besu.ethereum.api.handlers.AbstractJsonRpcExecutor.executeRequest(AbstractJsonRpcExecutor.java:87)
	at org.hyperledger.besu.ethereum.api.handlers.JsonRpcObjectExecutor.execute(JsonRpcObjectExecutor.java:57)
	at org.hyperledger.besu.ethereum.api.handlers.JsonRpcExecutorHandler.lambda$handler$0(JsonRpcExecutorHandler.java:57)
	at java.base/java.util.Optional.ifPresentOrElse(Optional.java:196)
	at org.hyperledger.besu.ethereum.api.handlers.JsonRpcExecutorHandler.lambda$handler$2(JsonRpcExecutorHandler.java:54)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1623)
```

### Versions (Add all that apply)
* Software version: besu/v23.4.4/osx-aarch_64/openjdk-java-20

### PR DESCRIPTION
**Pull request by alyokaz**
Enable a limit on the range of blocks that can be supplied to the JSON-RPC trace_filter method.

The limit has a default value and can be overridden with a command line option at start up.

fixes #5971

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 1]
```diff
@@ -1259,6 +1259,12 @@ static class PermissionsOptionGroup {
       description = "Specifies the number of last blocks to cache  (default: ${DEFAULT-VALUE})")
   private final Integer numberOfblocksToCache = 0;
 
+  @Option(
+      names = {"--rpc-max-trace-range"},
+      description =
+          "Specifies the maximum number of blocks for the trace_filter method (default: $DEFAULT-VALUE)")
```
Comment (Gabriel-Trintinalia) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1265]:
```suggestion
          "Specifies the maximum number of blocks for the trace_filter method. Must be >=0. 0 specifies no limit (default: $DEFAULT-VALUE)")
```

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 2]
```diff
@@ -1259,6 +1259,12 @@ static class PermissionsOptionGroup {
       description = "Specifies the number of last blocks to cache  (default: ${DEFAULT-VALUE})")
   private final Integer numberOfblocksToCache = 0;
 
+  @Option(
+      names = {"--rpc-max-trace-range"},
```
Comment (Gabriel-Trintinalia) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1263]: I think we can renamed it to `--rpc-max-trace-filter-range` as it only applies for this method.

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 3]
```diff
@@ -2526,7 +2532,8 @@ private ApiConfiguration apiConfiguration() {
             .gasPriceMax(apiGasPriceMax)
             .maxLogsRange(rpcMaxLogsRange)
             .gasCap(rpcGasCap)
-            .isGasAndPriorityFeeLimitingEnabled(apiGasAndPriorityFeeLimitingEnabled);
+            .isGasAndPriorityFeeLimitingEnabled(apiGasAndPriorityFeeLimitingEnabled)
+            .maxTraceRange(maxTraceRange);
```
Comment (Gabriel-Trintinalia) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:2536]: same here, maybe maxTraceFilterRange

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilter.java [hunk 4]
```diff
@@ -88,6 +93,19 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
...
+    } catch (IllegalArgumentException ex) {
+      LOG.atDebug()
+          .setMessage("eth_getLogs request {} failed:")
+          .addArgument(requestContext.getRequest())
+          .setCause(ex.getCause())
+          .log();
+      return new JsonRpcErrorResponse(
+          requestContext.getRequest().getId(), RpcErrorType.EXCEEDS_RPC_MAX_BLOCK_RANGE);
+    }
+
```
Comment (Gabriel-Trintinalia) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilter.java:108]: we could remove the try-catch here and use a `if` instead

File: besu/src/test/resources/everything_config.toml [hunk 5]
```diff
@@ -90,6 +90,7 @@ rpc-max-logs-range=100
 json-pretty-print-enabled=false
 cache-last-blocks=512
 rpc-gas-cap = 50000000
+rpc-max-trace-range=100
```
Comment (Gabriel-Trintinalia) [inline besu/src/test/resources/everything_config.toml:93]: same as above

**Review by Gabriel-Trintinalia — COMMENTED** [review]
Some comments in the code. 

- This is also a breaking change. Could we add a entry to the changelog? 
- Also, a unit test similar to `shouldFailIfParamsExceedMaxRange` would be nice

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilter.java [hunk 6]
```diff
@@ -88,6 +93,19 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
     final long toBlock = resolveBlockNumber(filterParameter.getToBlock());
     LOG.trace("Received RPC rpcName={} fromBlock={} toBlock={}", getName(), fromBlock, toBlock);
 
+    try {
+      if (maxRange > 0 && toBlock - fromBlock > maxRange)
+        throw new IllegalArgumentException(RpcErrorType.EXCEEDS_RPC_MAX_BLOCK_RANGE.getMessage());
+    } catch (IllegalArgumentException ex) {
+      LOG.atDebug()
+          .setMessage("eth_getLogs request {} failed:")
```
[the remark this answers is not in the document]
Comment (alyokaz) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilter.java]: Done

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 1 above)
Comment (alyokaz) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: Done

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 2 above)
Comment (alyokaz) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: Done

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 3 above)
Comment (alyokaz) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: Done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilter.java (hunk 4 above)
Comment (alyokaz) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilter.java]: Done

File: besu/src/test/resources/everything_config.toml (hunk 5 above)
Comment (alyokaz) [reply besu/src/test/resources/everything_config.toml]: Done

**Comment by alyokaz** [comment]
> Thank you for the PR! I have added some comments in the code.
> 
> * This is a breaking change. Could you please add a entry to the changelog?
> * Also, a unit test similar to `shouldFailIfParamsExceedMaxRange` would be nice

Changelog entry added.

I've added a unit test for TraceFilter covering the new option. I wanted to add a test for the `0` parameter, but without mocking `traceFilterWithPipeline` using something like Powermock, it would require extensive refactoring and I felt it might run orthogonal to the "big function" style of the code. I think maybe it might be better dealt with in an integration test?

**Review by Gabriel-Trintinalia — COMMENTED** [review]
Looking good, just a text suggestion and the the right copyright for the test class

File: CHANGELOG.md [hunk 7]
```diff
@@ -3,6 +3,7 @@
 ## 24.1.2-SNAPSHOT
 
 ### Breaking Changes
+- A Limit on the range of blocks passed to the JSON-RPC API `trace-filter` method, that defaults to 1000 and can be set with `--rpc-max-logs-range`, has been enabled [#6446](https://github.com/hyperledger/besu/pull/6446)
```
[the remark this answers is not in the document]
Comment (alyokaz) [reply CHANGELOG.md]: Changed as suggested

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilterTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
+/* SPDX-License-Identifier: Apache-2.0 */
```
[the remark this answers is not in the document]
Comment (alyokaz) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilterTest.java]: Added the copyright as it appears at the top of EthGetLogsTest

File: CHANGELOG.md [hunk 9]
```diff
@@ -3,6 +3,7 @@
 ## 24.1.2-SNAPSHOT
 
 ### Breaking Changes
+- The `trace-filter` method in JSON-RPC API now has a default block range limit of 1000, adjustable with `--rpc-max-logs-range` [#6446](https://github.com/hyperledger/besu/pull/6446)
```
Comment (macfarla) [inline CHANGELOG.md:6]:
```suggestion
- The `trace-filter` method in JSON-RPC API now has a default block range limit of 1000, adjustable with `--rpc-max-trace-filter-range` [#6446](https://github.com/hyperledger/besu/pull/6446)
```

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilterTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright ConsenSys AG.
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceFilterTest.java:2]:
```suggestion
 * Copyright Hyperledger Besu Contributors.
```

**Review by macfarla — APPROVED** [review]
approved with a couple of minor nits

**Pull request merged**